### PR TITLE
Base PR: Totems

### PR DESCRIPTION
**Pull request by jjpark-kb**
# What this PR does
adds totemic tomes
adds totems with different effects:
1) heal health allies
2) hurt health enemies
3) heal stamina allies
4) hurt stamina enemies
5) heal mana allies
6) hurt mana enemies
7) burn enemies
8) repel enemies
9) attract enemies

# Why it should be added to the game
continue adding magic content, especially a supporting mage class

### CODE REVIEW

**Comment by jjpark-kb** [comment]
I think that is enough content for a base PR

File: code/_core/obj/item/weapon/ranged/magic/tomes/summon/totem/_totem.dm [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		summoned_totem.totem_remove_time = duration + world.time
+		summoned_totem.leveled_effect = round(livingOwner.get_skill_level(SKILL_MAGIC)/50) //send help
+		if(livingOwner.is_player_controlled())
+			livingOwner.add_skill_xp(SKILL_MAGIC,cost_mana*1.5)
+	INITIALIZE(summoned_totem)
+	GENERATE(summoned_totem)
+	FINALIZE(summoned_totem)
+	if(summoned_totem)
+		summoned_totem.Move(get_turf(hit_atom))
+	return ..()
```
Comment (BurgerLUA) [inline code/_core/obj/item/weapon/ranged/magic/tomes/summon/totem/_totem.dm:40]: This should just be ``. = ..()``

File: code/_core/obj/structure/totem.dm [hunk 2]
```diff
@@ -0,0 +1,243 @@
+/obj/structure/totem
+	name = "totem"
+	desc = "None."
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:3]: A short description should be added for each of the totem types.

File: code/_core/obj/structure/totem.dm [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+	var/leveled_effect = 0
+
+	var/affecting_faction //which faction it should affect
+
+/obj/structure/totem/Finalize()
+	. = ..()
+	start_thinking(src)
+
+/obj/structure/totem/Destroy()
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:21]: This should also contain ``owner = null`` after ``owner.totem = null``

File: code/_core/mob/living/_living.dm [hunk 4]
```diff
@@ -201,6 +201,8 @@
 	var/mob/living/master //This object's master.
 	var/minion_remove_time = 0
 
+	var/obj/structure/totem/totem //This object's totem.
```
Comment (BurgerLUA) [inline code/_core/mob/living/_living.dm:204]: ``QDEL_NULL(totem)`` should be called in ``mob/living/Destroy()`` somewhere

File: code/_core/obj/structure/totem.dm [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+	if(owner)
+		owner.totem = null
+	return ..()
+
+/obj/structure/totem/think()
+	if(world.time >= totem_remove_time)
+		Destroy()
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:28]: ``Destroy()`` should not directly be called. ``qdel(src)`` should be called instead.

File: code/_core/obj/structure/totem.dm [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+			continue
+		if(L.health.health_current >= L.health.health_max)
+			continue
+		if(L.health.get_loss(BRUTE))
+			L.brute_regen_buffer += (3 + (3 * leveled_effect))
+		if(L.health.get_loss(BURN))
+			L.burn_regen_buffer += (3 + (3 * leveled_effect))
+		if(L.health.get_loss(TOX))
+			L.tox_regen_buffer += (3 + (3 * leveled_effect))
+		var/obj/effect/temp/healing/H = new(L.loc,10,COLOR_RED)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:65]: You should use ``CREATE(/obj/effect/temp/healing/,get_turf(L))`` instead and remove the ``INITIALIZE(H)`` and friends.

File: code/_core/obj/structure/totem.dm [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+		if(L.dead)
+			continue
+		if(L.loyalty_tag == affecting_faction) //== because we dont want to affect allies
+			continue
+		if(!istype(L.health))
+			continue
+		L.brute_regen_buffer -= (3 + (3 * leveled_effect))
+		L.burn_regen_buffer -= (3 + (3 * leveled_effect))
+		L.tox_regen_buffer -= (3 + (3 * leveled_effect))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_RED)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:87]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+			continue
+		if(L.loyalty_tag != affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		if(L.health.stamina_current >= L.health.stamina_max)
+			continue
+		if(L.health.get_stamina_loss())
+			L.stamina_regen_buffer += (3 + (3 * leveled_effect))
+		var/obj/effect/temp/healing/H = new(L.loc,10,COLOR_GREEN)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:110]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+	var/turf/T = get_turf(src)
+	for(var/mob/living/L in viewers(4,T))
+		if(L.dead)
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.stamina_regen_buffer -= (3 + (3 * leveled_effect))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_GREEN)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:130]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+			continue
+		if(L.loyalty_tag != affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		if(L.health.mana_current >= L.health.mana_max)
+			continue
+		if(L.health.get_mana_loss())
+			L.mana_regen_buffer += (3 + (3 * leveled_effect))
+		var/obj/effect/temp/healing/H = new(L.loc,10,COLOR_BLUE)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:153]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+	var/turf/T = get_turf(src)
+	for(var/mob/living/L in viewers(4,T))
+		if(L.dead)
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.mana_regen_buffer -= (3 + (3 * leveled_effect))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_BLUE)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:173]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+	var/turf/T = get_turf(src)
+	for(var/mob/living/L in viewers(4,T))
+		if(L.dead)
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.ignite(SECONDS_TO_DECISECONDS(1.5))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_RED)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:194]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+		if(L.dead)
+			continue
+		if(L.boss) //yea, no repelling bosses, sorry
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.Move(get_step(L.loc,get_dir(src,L)))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_BLUE)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:217]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+	for(var/mob/living/L in viewers(4,T))
+		if(L.dead)
+			continue
+		if(L.boss) //yea, no repelling bosses, sorry
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.Move(get_step(L.loc,get_dir(src,L)))
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:216]: A little nervous about calling Move directly like this but we'll have to see if this causes any exploits.

File: code/_core/obj/structure/totem.dm [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+		if(L.dead)
+			continue
+		if(L.boss) //yea, no attracting bosses, sorry
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.Move(get_step(L.loc,get_dir(L,src)))
+		var/obj/effect/temp/electricity/H = new(L.loc,10,COLOR_BLUE)
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:240]: You should use CREATE(/obj/effect/temp/healing/,get_turf(L)) instead and remove the INITIALIZE(H) and friends.

File: code/_core/obj/structure/totem.dm [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+	for(var/mob/living/L in viewers(4,T))
+		if(L.dead)
+			continue
+		if(L.boss) //yea, no attracting bosses, sorry
+			continue
+		if(L.loyalty_tag == affecting_faction)
+			continue
+		if(!istype(L.health))
+			continue
+		L.Move(get_step(L.loc,get_dir(L,src)))
```
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:239]: There might be an issue where someone might be on top of the totem and ``Move`` will be a redundant call. You should compare the two locs together to see if someone should actually move.

File: code/_core/obj/structure/totem.dm (hunk 14 above)
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:216]: Also, you should call ``if(!isturf(L.loc) continue`` before the movement in cases where the living target might actually be in something else, like a crate or a vehicle.

File: code/_core/obj/structure/totem.dm (hunk 16 above)
Comment (BurgerLUA) [inline code/_core/obj/structure/totem.dm:239]: Also, you should call ``if(!isturf(L.loc) continue`` before the movement in cases where the living target might actually be in something else, like a crate or a vehicle.

**Review by BurgerLUA — CHANGES_REQUESTED** [review]

**Comment by BurgerLUA** [comment]
Very good PR, by the way.

**Comment by jjpark-kb** [comment]
I tested it and it worked before I did the changes.
I HAVE NOT tested it just yet with these changes.

**Comment by jjpark-kb** [comment]
okay, tested. Works.

**Comment by jjpark-kb** [comment]
the longer this PR is up, the more I hunger to create more totems 👀

File: code/_core/mob/living/_living.dm [hunk 17]
```diff
@@ -283,6 +283,10 @@
 		master.minion = null
 		master = null
 
+	if(totem)
+		QDEL_NULL(totem)
+		totem = null
```
Comment (BurgerLUA) [inline code/_core/mob/living/_living.dm:288]: Don't need the null here but I can remove this when merging.

**Review by BurgerLUA — APPROVED** [review]

**Pull request merged**